udmis/src/test/java/com/google/bos/udmi/monitoring/LogTailTest.java

```diff
@@ -0,0 +1,329 @@
+    Mockito.doReturn(streamMock).when(loggingMock).tailLogEntries(any(TailOption.class));
+
+    assertEquals(streamMock, logTailMock.getCloudLogStream(null));
+  }
+
+  @Test
+  public void testMain() throws org.apache.commons.cli.ParseException {
+    LogTail logTailMock = getLogTailMock();
+    Mockito.doNothing().when(logTailMock).tailLogs();
+    LogTail.mainUnderTest(TEST_MAIN_ARGS, logTailMock);
+  }
+
+  @Test
+  public void testMainNoProject() {
+    String[] args = {};
+    assertThrows(RuntimeException.class, () -> LogTail.main(args));
+  }
+
+  @Test
+  public void testProcessLogEntryError() {
+    LogTail logTailMock = getLogTailMock();
+
+    LogEntry logEntryMock = Mockito.mock(LogEntry.class);
+    ProtoPayload payloadMock = Mockito.mock(ProtoPayload.class);
+    Any dataMock = Mockito.mock(Any.class);
+    ByteString byteStringMock = Mockito.mock(ByteString.class);
+    byte[] byteArray = {0};
+
+    Mockito.when(logEntryMock.getPayload()).thenReturn(payloadMock);
+    Mockito.when(payloadMock.getData()).thenReturn(dataMock);
+    Mockito.when(dataMock.getTypeUrl())
+        .thenReturn("type.googleapis.com/google.cloud.audit.AuditLog");
+    Mockito.when(dataMock.getValue()).thenReturn(byteStringMock);
+    Mockito.when(byteStringMock.toByteArray()).thenReturn(byteArray);
+
+    AuditLog auditLogMock = Mockito.mock(AuditLog.class);
+    try {
+      Mockito.doReturn(auditLogMock).when(logTailMock).getAuditLogParseFrom(any(byte[].class));
+    } catch (InvalidProtocolBufferException e) {
+      fail(e);
+    }
+
+    LogTailEntry entryTest = getLogTailEntryTest();
+    Mockito.doReturn(entryTest.timestamp).when(logEntryMock).getInstantTimestamp();
+    Mockito.doReturn(entryTest.methodName).when(auditLogMock).getMethodName();
+    Mockito.doReturn(entryTest.serviceName).when(auditLogMock).getServiceName();
+    Mockito.doReturn(entryTest.resourceName).when(auditLogMock).getResourceName();
+    Status statusMock = Mockito.mock(Status.class);
+    Mockito.doReturn(statusMock).when(auditLogMock).getStatus();
+    Mockito.doReturn(entryTest.statusCode).when(statusMock).getCode();
+    Mockito.doReturn(entryTest.statusMessage).when(statusMock).getMessage();
+    Mockito.doReturn(entryTest.severity).when(logEntryMock).getSeverity();
+
+    logTailMock.logsTimeSeries = Mockito.mock(LogTimeSeries.class);
+    ArgumentCaptor<LogTailEntry> argumentCaptor = ArgumentCaptor.forClass(LogTailEntry.class);
+    Mockito.doNothing().when(logTailMock.logsTimeSeries).add(argumentCaptor.capture());
+    try {
+      Mockito.doNothing().when(logTailMock.logsTimeSeries)
+          .maybeEmitMetrics(any(LogTailOutput.class));
+    } catch (IOException e) {
+      fail(e);
+    }
+
+    logTailMock.processLogEntryError(logEntryMock);
+    assertTrue(argumentCaptor.getValue().equals(entryTest));
+  }
+
+  /*
+   This test fails because IOException is not thrown from maybeEmitMetrics().
```

Comment:
but what's the problem?  Is it that it should throw an exception, or that the test is wrong?

Reply:
(I sent this to generate your next round of comments in parallel of working on this) Yeah the test doesn't function -- the test code does not function so that the code UNDER test throws a IOException from maybeEmitMetrics.